src/compiler/commandLineParser.ts

```diff
@@ -884,13 +888,306 @@ namespace ts {
+                    continue;
+                }
+
+                if (element.questionToken) {
+                    errors.push(createDiagnosticForNodeInSourceFile(sourceFile, element.questionToken, Diagnostics._0_can_only_be_used_in_a_ts_file, "?"));
+                }
+                if (!isDoubleQuotedString(element.name)) {
+                    errors.push(createDiagnosticForNodeInSourceFile(sourceFile, element.name, Diagnostics.String_literal_with_double_quotes_expected));
+                }
+
+                const keyText = getTextOfPropertyName(element.name);
+                const option = knownOptions ? knownOptions.get(keyText) : undefined;
+                if (extraKeyDiagnosticMessage && !option) {
+                    errors.push(createDiagnosticForNodeInSourceFile(sourceFile, element.name, extraKeyDiagnosticMessage, keyText));
+                }
+                const value = convertPropertyValueToJson(element.initializer, option);
+                if (typeof keyText !== undefined && typeof value !== undefined) {
+                    result[keyText] = value;
+                    // Notify key value set, if user asked for it
+                    if (jsonConversionNotifier &&
+                        // Current callbacks are only on known parent option or if we are setting values in the root
+                        (parentOption || knownOptions === knownRootOptions)) {
+                        const isValidOptionValue = isCompilerOptionsValue(option, value);
+                        if (parentOption) {
+                            if (isValidOptionValue) {
+                                // Notify option set in the parent if its a valid option value
+                                jsonConversionNotifier.onSetValidOptionKeyValueInParent(parentOption, option, value);
+                            }
+                        }
+                        else if (knownOptions === knownRootOptions) {
+                            if (isValidOptionValue) {
+                                // Notify about the valid root key value being set
+                                jsonConversionNotifier.onSetValidOptionKeyValueInRoot(keyText, element.name, value, element.initializer);
+                            }
+                            else if (!option) {
+                                // Notify about the unknown root key value being set
+                                jsonConversionNotifier.onSetUnknownOptionKeyValueInRoot(keyText, element.name, value, element.initializer);
+                            }
+                        }
+                    }
+                }
+            }
+            return result;
+        }
+
+        function convertArrayLiteralExpressionToJson(
+            elements: NodeArray<Expression>,
+            elementOption: CommandLineOption | undefined
+        ): any[] {
+            const result: any[] = [];

```

Comment:
This could just use `elements.map`? If you also want to filter out undefined values, maybe `mapDefined`.